core/os/os2/process_linux.odin

```diff
@@ -71,25 +302,446 @@ _Sys_Process_Attributes :: struct {}
+	}
+	defer if desc.working_dir != "" {
+		linux.close(dir_fd)
+	}
+
+	// search PATH if just a plain name is provided
+	exe_fd: linux.Fd
+	executable_name := desc.command[0]
+	if strings.index_byte(executable_name, '/') == -1 {
+		path_env := get_env("PATH", temp_allocator())
+		path_dirs := filepath.split_list(path_env, temp_allocator())
+
+		found: bool
+		for dir in path_dirs {
+			exe_path := fmt.caprintf("%s/%s", dir, executable_name, allocator=temp_allocator())
+			if exe_fd, errno = linux.openat(dir_fd, exe_path, {.PATH, .CLOEXEC}); errno != .NONE {
+				continue
+			}
+			if !has_executable_permissions(exe_fd) {
+				linux.close(exe_fd)
+				continue
+			}
+			found = true
+			break
+		}
+		if !found {
+			// check in cwd to match windows behavior
+			exe_path := fmt.caprintf("./%s", executable_name, allocator=temp_allocator())
+			if exe_fd, errno = linux.openat(dir_fd, exe_path, {.PATH, .CLOEXEC}); errno != .NONE {
+				return process, .Not_Exist
+			}
+			if !has_executable_permissions(exe_fd) {
+				linux.close(exe_fd)
+				return process, .Permission_Denied
+			}
+		}
+	} else {
+		exe_path := temp_cstring(executable_name) or_return
+		if exe_fd, errno = linux.openat(dir_fd, exe_path, {.PATH, .CLOEXEC}); errno != .NONE {
+			return process, _get_platform_error(errno)
+		}
+		if !has_executable_permissions(exe_fd) {
+			linux.close(exe_fd)
+			return process, .Permission_Denied
+		}
+	}
+
+	// At this point, we have an executable.
+	defer linux.close(exe_fd)
+
+	// args and environment need to be a list of cstrings
+	// that are terminated by a nil pointer.
+	cargs := make([]cstring, len(desc.command) + 1, temp_allocator())
+	for command, i in desc.command {
+		cargs[i] = temp_cstring(command) or_return
+	}
+
+	// Use current process' environment if description didn't provide it.
+	env: [^]cstring
+	if desc.env == nil {
+		// take this process's current environment
+		env = raw_data(export_cstring_environment(temp_allocator()))
+	} else {
+		cenv := make([]cstring, len(desc.env) + 1, temp_allocator())
+		for env, i in desc.env {
+			cenv[i] = temp_cstring(env) or_return
+		}
+		env = &cenv[0]
+	}
+
+	child_pipe_fds: [2]linux.Fd
+	if errno = linux.pipe2(&child_pipe_fds, {.CLOEXEC}); errno != .NONE {
+		return process, _get_platform_error(errno)
+	}
+
+	// TODO: This is the traditional textbook implementation with fork.
+	//       A more efficient implementation with vfork:
+	//
+	//       1. retrieve signal handlers
+	//       2. block all signals
+	//       3. allocate some stack space
+	//       4. vfork (waits for child exit or execve); In child:
+	//           a. set child signal handlers
+	//           b. set up any necessary pipes
+	//           c. execve
+	//       5. restore signal handlers
+	//
+	pid: linux.Pid
+	if pid, errno = linux.fork(); errno != .NONE {
+		return process, _get_platform_error(errno)
+	}
+
+	STDIN  :: linux.Fd(0)
+	STDOUT :: linux.Fd(1)
+	STDERR :: linux.Fd(2)
+
+	READ :: 0
+	WRITE :: 1
+
+	if pid == 0 {
```

Comment:
Should the child not close the read side of the pipe here? `linux.close(child_pipe_fds[READ])`

Reply:
That pipe is opened with `CLOEXEC`.  In the child, they will not exist passed the `execve`.